pkg/kubectl/cmd/rollout/rollout_status.go

```diff
@@ -158,60 +181,70 @@ func (o *RolloutStatusOptions) Run() error {
-		return err
-	}
-	fmt.Fprintf(o.Out, "%s", status)
-	if done {
-		return nil
+	fieldSelector := fields.OneTermEqualSelector("metadata.name", info.Name).String()
+	lw := &cache.ListWatch{
+		ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
+			options.FieldSelector = fieldSelector
+			return o.DynamicClient.Resource(info.Mapping.Resource).Namespace(info.Namespace).List(options)
+		},
+		WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+			options.FieldSelector = fieldSelector
+			return o.DynamicClient.Resource(info.Mapping.Resource).Namespace(info.Namespace).Watch(options)
+		},
 	}
 
-	shouldWatch := o.Watch
-	if !shouldWatch {
-		return nil
-	}
+	preconditionFunc := func(store cache.Store) (bool, error) {
+		_, exists, err := store.Get(&metav1.ObjectMeta{Namespace: info.Namespace, Name: info.Name})
+		if err != nil {
+			return true, err
+		}
+		if !exists {
+			// We need to make sure we see the object in the cache before we start waiting for events
+			// or we would be waiting for the timeout if such object didn't exist.
+			return true, apierrors.NewNotFound(mapping.Resource.GroupResource(), info.Name)
+		}
 
-	// watch for changes to the deployment
-	w, err := r.Watch(rv)
-	if err != nil {
-		return err
+		return false, nil
 	}
 
 	// if the rollout isn't done yet, keep watching deployment status
-	// TODO: expose timeout
-	timeout := 0 * time.Second
-	ctx, cancel := watchtools.ContextWithOptionalTimeout(context.Background(), timeout)
-	defer cancel()
+	ctx, cancel := watchtools.ContextWithOptionalTimeout(context.Background(), o.Timeout)
 	intr := interrupt.New(nil, cancel)
 	return intr.Run(func() error {
-		_, err := watchtools.UntilWithoutRetry(ctx, w, func(e watch.Event) (bool, error) {
-			// print deployment's status
-			status, done, err := statusViewer.Status(info.Namespace, info.Name, o.Revision)
-			if err != nil {
-				return false, err
-			}
-			fmt.Fprintf(o.Out, "%s", status)
-			// Quit waiting if the rollout is done
-			if done {
-				return true, nil
+		_, err = watchtools.UntilWithSync(ctx, lw, &unstructured.Unstructured{}, preconditionFunc, func(e watch.Event) (bool, error) {
+			switch t := e.Type; t {
+			case watch.Added, watch.Modified:
+				status, done, err := statusViewer.Status(e.Object.(runtime.Unstructured), o.Revision)
+				if err != nil {
+					return false, err
+				}
+				fmt.Fprintf(o.Out, "%s", status)
+				// Quit waiting if the rollout is done
+				if done {
+					return true, nil
+				}
+
+				shouldWatch := o.Watch
+				if !shouldWatch {
+					return true, nil
+				}
+
+				return false, nil
+
+			case watch.Deleted:
+				// We need to abort to avoid cases of recreation and not to silently watch the wrong (new) object
+				return true, fmt.Errorf("object has been deleted")
+
+			default:
+				glog.Fatalf("unexpected event %#v", e)
+				// Compiler doesn't see that glog.Fatalf exits the flow, panic makes it happy.
+				panic("glog failed to exit")
```

Comment:
Don't, return empty error instead.

Reply:
Same here, if glog.Fatal were not to exit something is horribly wrong. There might be other ways than panic to deal with this but we need to error out here, not ignore that the universe is falling apart :) I personally think the panic is appropriate action if such state should ever be reached.

Reply:
Having `V(1)` error and checking for its not appearing in tests is the same approach. Although I'm not happy with us having `panic`-s in code :/

Reply:
If you want to go with panic, do `panic(fmt.Sprintf(unexpected event %#v", e))` instead.

Reply:
Bypassing compiler problems with double lines is wrong.

Reply:
> Having V(1) error and checking for its not appearing in tests is the same approach.

I don't see how that's comparable. This needs to end with the CLI command exiting with non 0 return code. I am open to other paths that end up in that place.

> Although I'm not happy with us having panic-s in code :/

In normal code paths this is correct. This is a bit special case since it's only run if we asked to exit the program and it didn't - seems like totally appropriate action to what happened.

> If you want to go with panic, do panic(fmt.Sprintf(unexpected event %#v", e)) instead.

this seemed resonable on the first look, but we are not panicing because unexpected event, but because our call to exit the program failed.

> Bypassing compiler problems with double lines is wrong.

It's not ideal, but much better then ignoring errors.

Reply:
wanted to file an issue in the glog to fix this; the I've found out:
```
The master copy of the source lives inside
Google, not here. The code in this repo is for export only and is not itself
under development. Feature requests will be ignored.
```
just awesome to be based on such internal project...

Reply:
went for regular error which will result in setting the exit code to non-zero as well. I believe this removes the objections you had for fatal/panic since they are now gone

/hold cancel